File: frontend/src/types/station.ts
import { Reading } from './reading';

export type StationStatus =
  | 'OK'
  | 'MONITOR'
  | 'SCHEDULE'
  | 'SERVICE NOW'
  | 'HEALTHY'
  | 'SERVICE_NOW'
  | 'NO_DATA';

export interface SensorHealthItem {
  sensor: string;
  status: 'HEALTHY' | 'DEGRADED' | 'FAILED' | 'UNKNOWN';
  current_value: number | null;
  unit: string;
  last_calibrated?: string;
  drift_score: number;
  flags: string[];
}

export interface StationMapPoint {
  id: string;
  name: string;
  code: string;
  latitude: number;
  longitude: number;
  status: StationStatus;
  health_score: number;
  current_temp?: number | null;
  current_pressure?: number | null;
  data_quality?: string | null;
  condition?: string | null;
  degradation?: number | null;
  trend_per_day?: number | null;
  days_to_threshold?: number | null;
  high_conf_alerts?: number;
  alert_rate_pct?: number;
  rate_vs_network?: number;
}

export interface StationHealthSummary {
  id: string;
  name: string;
  code: string;
  state: string;
  district: string;
  latitude: number;
  longitude: number;
  elevation_m: number;
  status: StationStatus;
  health_score: number;
  last_seen: string;
  active_alerts_count: number;
  sensors: SensorHealthItem[];
  degradation?: number | null;
  trend_per_day?: number | null;
  days_to_threshold?: number | null;
  high_conf_alerts?: number;
  alert_rate_pct?: number;
  rate_vs_network?: number;
}

export interface StationDetail extends StationHealthSummary {
  latest_reading?: Reading | null;
  neighbour_agreement_pct: number;
  why_flagged_summary?: string | null;
}

export interface PaginatedStationsResponse {
  items: StationHealthSummary[];
  total: number;
  page: number;
  pages: number;
  limit: number;
}

File: backend/app/schemas/station.py
from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, ConfigDict
from .reading import ReadingResponse

class SensorHealthItem(BaseModel):
    sensor: str # TEMPERATURE, HUMIDITY, PRESSURE, WIND, SOLAR, PRECIP
    status: str # HEALTHY, DEGRADED, FAILED, UNKNOWN
    current_value: Optional[float] = None
    unit: str
    last_calibrated: Optional[str] = None
    drift_score: float = 0.0
    flags: List[str] = []

class StationBase(BaseModel):
    name: str
    code: str
    state: str
    district: str
    latitude: float
    longitude: float
    elevation_m: float = 0.0

class StationCreate(StationBase):
    id: str

class StationMapPoint(BaseModel):
    id: str
    name: str
    code: str
    latitude: float
    longitude: float
    status: str
    health_score: float
    current_temp: Optional[float] = None
    current_pressure: Optional[float] = None

class StationHealthSummary(StationBase):
    id: str
    status: str # HEALTHY, MONITOR, SERVICE_NOW, NO_DATA
    health_score: float
    last_seen: datetime
    active_alerts_count: int = 0
    sensors: List[SensorHealthItem] = []

    model_config = ConfigDict(from_attributes=True)

class StationDetail(StationHealthSummary):
    latest_reading: Optional[ReadingResponse] = None
    neighbour_agreement_pct: float = 100.0
    why_flagged_summary: Optional[str] = None

class PaginatedStationsResponse(BaseModel):
    items: List[StationHealthSummary]
    total: int
    page: int
    pages: int
    limit: int
